cmd/juju/controller/register.go

```diff
@@ -98,34 +110,134 @@ func (c *registerCommand) Init(args []string) error {
+		controllerDetails,
+		accountDetails,
+	)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Log into the controller to verify the credentials, and
+	// list the models available.
+	models, err := c.listModelsFunc(store, controllerName, accountDetails.User)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, model := range models {
+		owner := names.NewUserTag(model.Owner)
+		if err := store.UpdateModel(
+			controllerName,
+			jujuclient.JoinOwnerModelName(owner, model.Name),
+			jujuclient.ModelDetails{model.UUID},
+		); err != nil {
+			return errors.Annotate(err, "storing model details")
+		}
+	}
+	if err := store.SetCurrentController(controllerName); err != nil {
 		return errors.Trace(err)
 	}
 
+	fmt.Fprintf(
+		ctx.Stderr, "\nWelcome, %s. You are now logged into %q.\n",
+		friendlyUserName(accountDetails.User), controllerName,
+	)
+	return c.maybeSetCurrentModel(ctx, store, controllerName, accountDetails.User, models)
+}
+
+func friendlyUserName(user string) string {
+	u := names.NewUserTag(user)
+	if u.IsLocal() {
+		return u.Name()
+	}
+	return u.Id()
+}
+
+// controllerDetails returns controller and account details to be registered for the
+// given registration parameters.
+func (c *registerCommand) controllerDetails(ctx *cmd.Context, p *registrationParams) (jujuclient.ControllerDetails, jujuclient.AccountDetails, error) {
+	if p.publicHost != "" {
+		return c.publicControllerDetails(p.publicHost)
+	}
+	return c.nonPublicControllerDetails(ctx, p)
+}
+
+// publicControllerDetails returns controller and account details to be registered
+// for the given public controller host name.
+func (c *registerCommand) publicControllerDetails(host string) (jujuclient.ControllerDetails, jujuclient.AccountDetails, error) {
+	errRet := func(err error) (jujuclient.ControllerDetails, jujuclient.AccountDetails, error) {
+		return jujuclient.ControllerDetails{}, jujuclient.AccountDetails{}, err
+	}
+	if !strings.Contains(host, ":") {
+		host += ":443"
+	}
+	// Make a direct API connection because we don't yet know the
+	// controller UUID so can't store the thus-incomplete controller
+	// details to make a conventional connection.
+	//
+	// Unfortunately this means we'll connect twice to the controller
+	// but it's probably best to go through the conventional path the
+	// second time.
+	bclient, err := c.BakeryClient()
+	if err != nil {
+		return errRet(errors.Trace(err))
+	}
+	dialOpts := api.DefaultDialOpts()
+	dialOpts.BakeryClient = bclient
+	conn, err := c.apiOpen(&api.Info{
+		Addrs: []string{host},
+	}, dialOpts)
+	if err != nil {
+		return errRet(errors.Trace(err))
+	}
+	defer conn.Close()
+	user, ok := conn.AuthTag().(names.UserTag)
+	if !ok {
+		return errRet(errors.Errorf("logged in as %v, not a user", conn.AuthTag()))
+	}
+	return jujuclient.ControllerDetails{
+			APIEndpoints:   []string{host},
+			ControllerUUID: conn.ControllerTag().Id(),
+			// TODO(rogpeppe) it might be good to update the non-official CA cert here too
```

Comment:
I've removed the TODO. I'm not sure it's the right thing to do anyway and - it doesn't justify a bug report. (BTW only 12% of TODOs in the source code fit this format. I'm not entirely convinced it's worth having an issue for every TODO).